### PR TITLE
Add serilog with default console logger.

### PR DESCRIPTION
**Pull request by joni7777**
Add implementation for mattermost logger and splunk logger
Activate loggers by configuration.

Fix github issues #1 #3

### CODE REVIEW

File: Bp.Logging/Bp.Logging.csproj [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>netcoreapp2.2</TargetFramework>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <Reference Include="Microsoft.AspNetCore.Hosting.Abstractions, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60">
```
Comment (Meir017) [inline Bp.Logging/Bp.Logging.csproj]: remove this

File: Bp.Logging/Sinks/Http/BpBulkHttpSink.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        public void Dispose()
+        {
+            _httpClient?.Dispose();
+            _timer?.Dispose();
+        }
+
+        private ConcurrentBag<LogEvent> EmptyLogsBag(ConcurrentBag<LogEvent> logs)
+        {
+            ConcurrentBag<LogEvent> existingLogs = new ConcurrentBag<LogEvent>();
```
Comment (Meir017) [inline Bp.Logging/Sinks/Http/BpBulkHttpSink.cs:61]: why not something as simple as
```cs
var existingLogs = logs.ToArray();
logs = new ConcurrentBag<LogEvent>();
return existingLogs;
```

File: Bp.Logging/Sinks/Splunk/BpSplunkSink.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        protected override string GenerateRequestUri(LogEvent logEvent)
+        {
+            logEvent.Properties.TryGetValue("ServiceName", out var serviceName);
+            return
+                $"https://{_splunkInfo.Host}:{_splunkInfo.Port}/services/receivers/simple?index={_splunkInfo.Index}&sourcetype={_splunkInfo.SourceType}&source={serviceName}&output_mode=json";
+        }
+
+        protected override HttpContent GenerateRequestBody(ConcurrentBag<LogEvent> logs)
+        {
+            var joinedLogs = "";
```
Comment (Meir017) [inline Bp.Logging/Sinks/Splunk/BpSplunkSink.cs]: use `StringBuilder` instead

File: Bp.Logging/Sinks/Splunk/BpSplunkSink.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+            foreach (LogEvent log in logs)
+            {
+                joinedLogs += string.Format("{0}{0}", _formater.Format(log), Environment.NewLine);
```
Comment (Meir017) [inline Bp.Logging/Sinks/Splunk/BpSplunkSink.cs]: this should probably be `string.Format("{0}{1}",` ...

File: Bp.Logging/Sinks/Http/BpBulkHttpSink.cs (hunk 2 above)
Comment (joni7777) [reply Bp.Logging/Sinks/Http/BpBulkHttpSink.cs:61]: I was afraid of context switch like this:

```
var existingLogs = logs.ToArray();
// CONTEXT SWTICH (Logs added to logs variable)
// Some logs will be gone
logs = new ConcurrentBag<LogEvent>();
return existingLogs;
```

File: Bp.Logging/SerilogInit.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+{
+    public static class SerilogInit
+    {
+        public static void ConfigureLogger(WebHostBuilderContext builder, LoggerConfiguration configuration)
+        {
+            configuration
+                .ReadFrom.Configuration(builder.Configuration)
+                .Enrich.WithProperty("ServiceName", builder.Configuration.GetSection("Service")["name"]);
+
+            var splunkLogger = builder.Configuration.GetSection("SplunkLogger");
```
Comment (Meir017) [inline Bp.Logging/SerilogInit.cs]: add `.Get<BpSplunkInfo>();` (`BpSplunkInfo` must contain a parameterless ctor and all properties should be `{ get; set; }`)

add package `Microsoft.Extensions.Configuration.Binder`

File: Bp.ApiRunner/Startup.cs [hunk 6]
```diff
@@ -54,7 +57,7 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
             app.UseBpHealthChecks();
             app.UseMvc();
             app.UseSwagger();
-            app.UseSwaggerUI(c => { c.SwaggerEndpoint("/swagger/v1/swagger.json", "Bp Seed V1"); });
+            app.UseSwaggerUI(c => { c.SwaggerEndpoint($"/swagger/{_service.Version}/swagger.json", $"{_service.Name} V{_service.Version}"); });
```
Comment (Meir017) [inline Bp.ApiRunner/Startup.cs]: add property `SwaggerUrl` for this

File: Bp.Logging/Sinks/Http/BpBulkHttpSink.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            
+            while (logs.Count > 0)
```
Comment (Meir017) [inline Bp.Logging/Sinks/Http/BpBulkHttpSink.cs]: use `.IsEmpty`

File: Bp.Logging/Sinks/Mattermost/BpMattermostSinkExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+using Serilog.Configuration;
+using Serilog.Events;
+
+namespace Bp.Logging.Sinks.Mattermost
+{
+    public static class BpMattermostSinkExtensions
+    {
+        public static LoggerConfiguration BpMattermostSink(
+            this LoggerSinkConfiguration loggerConfiguration,
+            BpMattermostInfo bpMattermostInfo, ConcurrentBag<LogEvent> logs = null, Timer timer = null,
```
Comment (Meir017) [inline Bp.Logging/Sinks/Mattermost/BpMattermostSinkExtensions.cs]: remove unused arguments

File: Bp.Logging/Sinks/Splunk/BpSplunkLogFormatter.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Bp.Logging.Sinks.Splunk
+{
+    public class BpSplunkLogFormatter
+    {
+        public static string Format(LogEvent logEvent)
+        {
+            var formattedLog = new StringBuilder();
+
+            formattedLog.Append($"timestamp=\"{DateTime.Now.Millisecond.ToString()}\" ");
```
Comment (Meir017) [inline Bp.Logging/Sinks/Splunk/BpSplunkLogFormatter.cs]: use `TotalMilliseconds`

File: Bp.Logging/Sinks/Splunk/BpSplunkLogFormatter.cs [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+            formattedLog.Append($"severity=\"{logEvent.Level}\" ");
+            formattedLog.Append($"action=\"{logEvent.RenderMessage()}\" ");
+
+            foreach (KeyValuePair<string, LogEventPropertyValue> property in logEvent.Properties)
```
Comment (Meir017) [inline Bp.Logging/Sinks/Splunk/BpSplunkLogFormatter.cs]: use `var`

File: Bp.Logging/Sinks/Splunk/BpSplunkSink.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+using System.Threading;
+using Bp.Logging.Sinks.Http;
+using Serilog.Events;
+
+namespace Bp.Logging.Sinks.Splunk
+{
+    public class BpSplunkSink : BpBulkHttpSink
+    {
+        private readonly BpSplunkInfo _splunkInfo;
+        private AuthenticationHeaderValue _basicAuth;
```
Comment (Meir017) [inline Bp.Logging/Sinks/Splunk/BpSplunkSink.cs]: make readonly

File: Components/Config/appsettings.json [hunk 12]
```diff
@@ -1,8 +1,46 @@
...
+      "Serilog.Sinks.Console"
+    ],
+    "MinimumLevel": {
+      "Default": "Debug",
+      "Override": {
+        "System": "Information",
+        "Microsoft": "Information"
+      }
+    },
+    "CustomLoggers":{
```
Comment (Meir017) [inline Components/Config/appsettings.json:10]: don't use serilogs magic configuration, all manual

File: Components/Sample/Implementations/SampleRouter.cs [hunk 13]
```diff
@@ -1,16 +1,27 @@
...
+using Microsoft.Extensions.Logging;
 
 namespace Components.Sample.Implementations
 {
     public class SampleRouter : RouterBase
     {
         private readonly ISampleController _sampleController = new SampleControllerFactory().Create();
+        private readonly ILogger _logger;
+
+        public SampleRouter([FromServices] ILogger<SampleRouter> logger)
```
Comment (Meir017) [inline Components/Sample/Implementations/SampleRouter.cs]: remove `FromServices`

**Review by Meir017 — CHANGES_REQUESTED** [review]

File: Bp.ApiRunner/Startup.cs [hunk 14]
```diff
@@ -54,7 +58,8 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
             app.UseBpHealthChecks();
             app.UseMvc();
             app.UseSwagger();
-            app.UseSwaggerUI(c => { c.SwaggerEndpoint("/swagger/v1/swagger.json", "Bp Seed V1"); });
+            app.UseSwaggerUI(c =>
+                c.SwaggerEndpoint($"/swagger/{_service.Version}/swagger.json", $"{_service.Name} V{_service.Version}"));
```
Comment (Meir017) [inline Bp.ApiRunner/Startup.cs]: create a property for this `$"/swagger/{_service.Version}/swagger.json"`

File: Bp.Logging/Sinks/Mattermost/BpMattermostInfo.cs [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+            !String.IsNullOrEmpty(Port) && !String.IsNullOrEmpty(Path);
+
+        public const string MATTERMOST_MISSING_CONFIGURATION_MESSAGE = "Mattermost missing required configuration";
+
+        public BpMattermostInfo()
+        {
+            
+        }
+
+        public BpMattermostInfo(string protocol, string host, string port, string path)
```
Comment (Meir017) [inline Bp.Logging/Sinks/Mattermost/BpMattermostInfo.cs]: remove this ctor

File: Bp.ApiRunner/ServiceInfo.cs [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+    public class ServiceInfo
+    {
+        public string Name { get; set; }
+        public string Version { get; set; }
+
+        public ServiceInfo()
+        {
+        }
+
+        public ServiceInfo(string name, string version)
```
Comment (Meir017) [inline Bp.ApiRunner/ServiceInfo.cs]: remove this ctor

File: Bp.Logging/Sinks/Splunk/BpSplunkInfo.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+            !String.IsNullOrEmpty(Protocol) &&!String.IsNullOrEmpty(Host) &&
+            !String.IsNullOrEmpty(Port) && !String.IsNullOrEmpty(Username) &&
+            !String.IsNullOrEmpty(Password) && !String.IsNullOrEmpty(Index) &&
+            !String.IsNullOrEmpty(SourceType);
+
+        public const string SPLUNK_MISSING_CONFIGURATION_MESSAGE = "Splunk missing required configuration";
+
+        public BpSplunkInfo(){}
+
+        public BpSplunkInfo(string protocol, string host, string port, string username, string password, string index, string sourceType)
```
Comment (Meir017) [inline Bp.Logging/Sinks/Splunk/BpSplunkInfo.cs]: remove this ctor

**Review by Meir017 — CHANGES_REQUESTED** [review]

File: Bp.ApiRunner/Startup.cs [hunk 18]
```diff
@@ -55,6 +56,7 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
                 app.UseHttpsRedirection();
             }
 
+            app.UseMiddleware<EnrichWithRequestParamsMiddleware>();
```
Comment (Meir017) [inline Bp.ApiRunner/Startup.cs]: create a static class with an extension method `UseEnrichWithRequestParams` in the same file as the `EnrichWithRequestParamsMiddleware` class

File: Bp.ApiRunner/Startup.cs [hunk 19]
```diff
@@ -68,5 +70,10 @@ public class SchoolContext : DbContext
         public SchoolContext(DbContextOptions<SchoolContext> options) : base(options)
         {
         }
+
```
Comment (Meir017) [inline Bp.ApiRunner/Startup.cs]: this looks like a sample class, remove it

File: Bp.Logging/EnrichWithRequestParams/EnrichWithRequestParamsMiddleware.cs [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        public Task InvokeAsync(HttpContext context)
+        {
+            LogContext.PushProperty("transactionId", context.TraceIdentifier);
+//            await context.Response.WriteAsync("------- Before ------ \n\r");
+//            await _next(context);
+
+//            await context.Response.WriteAsync("\n\r------- After ------");
+            return Task.CompletedTask;
```
Comment (Meir017) [inline Bp.Logging/EnrichWithRequestParams/EnrichWithRequestParamsMiddleware.cs]: need to call `_next` to call the next middleware in the pipeline

**Pull request merged**